src/coreclr/tools/Common/JitInterface/CorInfoImpl.cs

```diff
@@ -3430,6 +3430,7 @@ private void reportFatalError(CorJitResult result)
         {
             // We could add some logging here, but for now it's unnecessary.
             // CompileMethod is going to fail with this CorJitResult anyway.
+            _numFrameInfos = 0;
```

Comment:
Is this all state that needs to be cleaned up? From cursory look, there seems to be more state that may cause problems.

I am wondering whether the retry logic should be rather moved from the JIT to the EE side. It would allow the state cleanup to be more robust. Also, it would allow us to skip the retry in cases where it does not make sense. For example, it does not make sense to retry with no optimizations when generating Tier1 code.


@dotnet/jit-contrib Thoughts?

Reply:
> I am wondering whether the retry logic should be rather moved from the JIT to the EE side.

Maybe I'm missing something: the proposed change is in crossgen2, not the JIT (or EE).

In the JIT'ing case, we already have a backout function: `CEEJitInfo::BackoutJitData()` => ` EEJitManager::RemoveJitData()`. It removes EH info, GC info, and maybe more. (Maybe not the debug info?) Seems like crossgen2 would need to do the same (i.e., remove GC info, debug info, code buffer).

Reply:
We have similar method in crossgen2 as well (`CompileMethodCleanup`). The problem is that neither `BackoutJitData` nor `CompileMethodCleanup` gets called when the JIT decides to retry with optimizations off.

Reply:
Ah, I see. There's a "small" window where, after the JIT has already started allocating EH/GC/code space from the VM, where it can hit a NO_WAY assert and decide to retry. It would make sense for the JIT to call back through the JIT/EE interface before retrying saying "I'm going to retry".